tests/interfaces/test_tf.py

```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
+            qml.templates.StronglyEntanglingLayers(weights2, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        # input weights
+        weights1 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+        weights2 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=4)
+
+        weights1 = tf.Variable(weights1)
+        weights2 = tf.Variable(weights2)
+
+        # input data
+        data1 = tf.constant([0, 1, 1, 0], dtype=tf.float64) / np.sqrt(2)
+        data2 = tf.Variable([1, 1], dtype=tf.float64, trainable=False)
+
+        with tf.GradientTape() as tape:
+            loss = circuit(data1, weights1, data2, weights2)
+
+        grad = tape.gradient(loss, [weights1, weights2])
+        res1 = grad[0].numpy()
+        res2 = grad[1].numpy()
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert res1.shape == weights1.shape
+        assert res2.shape == weights2.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable elements of `weights`, not the data input
+        num_w1 = tf.size(weights1).numpy()
+        num_w2 = tf.size(weights2).numpy()
+
+        offset1 = tf.size(data1).numpy()
+        offset2 = tf.size(data1).numpy() + num_w1 + tf.size(data2).numpy()
+
+        expected = set(list(range(offset1, offset1 + num_w1)) + list(range(offset2, offset2 + num_w2)))
+        assert spy.call_args[1]["wrt"] == expected
+
+    def test_gradient_non_differentiable_none(self, mocker):
+        """Test that the gradient of a non-differentiable tensor is None"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="tf")
+        def circuit(weights):
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        weights = tf.constant(qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3))
+
+        with tf.GradientTape() as tape:
+            loss = circuit(weights)
+
+        grad = tape.gradient(loss, weights)
+        assert grad is None
+
+        spy.assert_not_called()
+
+    def test_non_differentiable_watch(self, mocker):
+        """Test that watching a non-differentiable tensor makes it differentiable"""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="tf")
+        def circuit(weights):
+            qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        weights = tf.constant(qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3))
+
+        with tf.GradientTape() as tape:
+            tape.watch([weights])
+            loss = circuit(weights)
+
+        grad = tape.gradient(loss, weights)
+        assert grad is not None
```

Comment:
Cool

Reply:
This was something I declared in the ADR was 'not possible', since it's not documented anywhere in the TF API 😆